profile_compilation/sandscout_compiler.py

```diff
@@ -39,335 +41,417 @@
+    # like a subpath.
+    # This might be too vague, but let's see what happens.
+    else:
+        # I am not appending a / since prefixes might state literals in the
+        # filter argument.
+        p[0] = "subpath(" + p[2] + ")"
+
+
+# TODO: this is sort of a hack and I should evaluate it effects carefully.
+# The issue is that requireNot can now accept entitlement requirements as
+# parameters.
 def p_simpleEntValObject(p):
-  'simpleEntValObject	: TK_REQENT TK_FILTER'
-  p[0] = p[1] +"("+ p[2] +")"
-  #p[0] = "todoFixNegatedEnts"
+    'simpleEntValObject	: TK_REQENT TK_FILTER'
+    p[0] = p[1] +"("+ p[2] +")"
+    #p[0] = "todoFixNegatedEnts"
+
 
 def p_otherType(p):
-  'otherType : TK_OTHERTYPE'
-  #Some filter types are capitalized, but this confuses Prolog, so we make them lowercase
-  p[1] = p[1].lower()
-  p[0] = p[1] 
- 
+    'otherType : TK_OTHERTYPE'
+    # Some filter types are capitalized, but this confuses Prolog, so we make
+    # them lowercase.
+    p[1] = p[1].lower()
+    p[0] = p[1]
+
+
 def p_regexFilter(p):
-  'regexFilter : TK_OTHERTYPE regexList'
-  p[0] = []
-  for r in p[2]:
-    r = r[1:]
-    r += "/i"
-    r = r.replace("\\.","[.]")
-    p[0].append( p[1] + "(" + r + ")" )
+    'regexFilter : TK_OTHERTYPE regexList'
+    p[0] = []
+    for r in p[2]:
+        if _CMD_ARGS:
+            r = r[1:]
+            r += "/i"
+            r = r.replace("\\.","[.]")
+
+        p[0].append( p[1] + "(" + r + ")" )
+
 
 def p_regexList(p):
-  '''regexList 	: TK_REGEXPRESSION regexList
-		| TK_REGEXPRESSION'''
-  if len(p) == 2:
-    p[0] = [p[1]]
-  if len(p) == 3:
-    p[0] = [p[1]] + p[2]
+    '''regexList 	: TK_REGEXPRESSION regexList
+        | TK_REGEXPRESSION'''
+    if len(p) == 2:
+        p[0] = [p[1]]
+    if len(p) == 3:
+        p[0] = [p[1]] + p[2]
+
 
 def p_error(p):
-  if p == None:
-    sys.stderr.write('line '+str(lexer.lineno)+': syntax error at EOF\n')
-  else:
-    sys.stderr.write('line '+str(lexer.lineno)+': syntax error near '+str(p.value)+'\n')
-  sys.exit()
+    if p == None:
+        sys.stderr.write(
+            'line ' + str(_LEXER.lineno) + ': syntax error at EOF\n')
+    else:
+        sys.stderr.write(
+            'line ' + str(_LEXER.lineno) + ': syntax error near '
+            + str(p.value) + '\n')
+    sys.exit()
+
+
+def _get_args():
+    global _CMD_ARGS
+
+    parser = ArgumentParser('Compile an Apple Sandbox profile to Prolog')
+    parser.add_argument('-i', '--input', dest='input_file', type=str,
+        required=True, help='the input file name')
+    parser.add_argument('-o', '--output', dest='output_file', type=str,
+        required=True, help='the output Prolog file name')
+
+    _CMD_ARGS = parser.parse_args()
+
+
+def parse_file(file_name):
+    input_file = open(file_name, "r").read()
+    _LEXER.input(input_file)
+    
+    parser = yacc.yacc()
+    parser.parse(input_file)
+
+    return _GRAPH
+
+
+def main():
+    _get_args()
```

Comment:
Why is this using and underscore prefix? Is there a Python naming convention for this?

Reply:
Yes. They're somewhat similar to `static` functions in C. If someone were to `from sandscout_compiler import *` [1], symbols starting with `_` would not be visible. However, when importing the file like this `import sandscout_compiler`, all symbols, regardless of names, would be imported.

[1] https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles